// js-deobfuscator/dist/modifications/renaming/nameMapping.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class NameMapping {
    /**
     * Creates a new name mapping.
     */
    constructor() {
        this.mappings = new Map();
        this.children = [];
    }
    /**
     * Adds a new mapping.
     * @param name The original variable name.
     * @param newName The new variable name.
     */
    addMapping(name, newName) {
        this.mappings.set(newName, name);
    }
    /**
     * Adds a child name mapping.
     * @param child The child name mapping.
     */
    addChild(child) {
        this.children.push(child);
    }
    /**
     * Serializes the mapping to an object (suitable for JSON.stringify).
     * @returns The mapping object.
     */
    serialize() {
        const data = {
            children: this.children.map(c => c.serialize())
        };
        for (const [newName, name] of this.mappings) {
            data[newName] = name;
        }
        return data;
    }
}
exports.default = NameMapping;
